arrow/arrow.py

```diff
@@ -614,15 +730,15 @@ def __getattr__(self, name):
             return int((self.month - 1) / self._MONTHS_PER_QUARTER) + 1
 
         if not name.startswith("_"):
-            value = getattr(self._datetime, name, None)
+            value: Optional[int] = getattr(self._datetime, name, None)
 
             if value is not None:
                 return value
 
-        return object.__getattribute__(self, name)
+        return cast(int, object.__getattribute__(self, name))
 
     @property
-    def tzinfo(self):
+    def tzinfo(self) -> Optional[dt_tzinfo]:
```

Comment:
We have `tzinfo` setter and it accepts `None` as tzinfo. So we can not assure that this method always returns non null value.

Reply:
We can restrict the setter also to prohibit `None`, but it will break backward compatibility.

Reply:
Yeah datetime allows for you to pass None, but I do not think Arrow itself ever sets tzinfo to None internally since we disallow naive datetime formation. I am pretty that in all cases where None is passed in, we substitute in `dateutil.tz.tzutc()` to prevent the creation of naive datetimes. CC @systemcatch to confirm this, but I am pretty confident on that.

Reply:
Hey @jadchaar, interestingly the `tzinfo` setter will cause a `None` value for tzinfo to occur.

```shell
(arrow) chris@ThinkPad:~/arrow$ python
Python 3.8.3 (default, Jul  7 2020, 18:57:36) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import arrow
>>> dt=arrow.utcnow()
>>> dt
<Arrow [2020-11-24T20:23:04.371964+00:00]>
>>> dt.tzinfo
tzutc()
>>> dt.tzinfo=None
>>> dt.tzinfo
>>> dt
<Arrow [2020-11-24T20:23:04.371964]>
```

I think we should prohibit this to avoid weird results. The best approach would be to remove the `tzinfo` setter entirely because that also helps us with `Arrow` objects being hashable yet also mutable.

Reply:
That's my point! Also I'm agree with removing `tzinfo` setter. `Arrow.replace()` have same functionality with `None` value filtering.

Reply:
But I think removing has to be discussed in another PR or issue.

Reply:
The setter is going to be removed in https://github.com/arrow-py/arrow/pull/902. Arrow objects should be immutable to prevent messing with hash behavior.

Reply:
Once the setter is removed, we should be able to change the return type from `Optional[dt_tzinfo]` to `dt_tzinfo` since Arrow objects are never naive.

Reply:
@jadchaar @systemcatch Changed it!